commands/eval.b.inc

```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * @file
+ * Command(s) for evaluating PHP code directly.
+ */
+
+/**
+ * Implements hook_b_command().
+ */
+function eval_b_command() {
+  return array(
+    'php-eval' => array(
+      'description' => bt('Evaluate arbitrary PHP code after bootstrapping Backdrop.'),
+      'callback' => 'php_eval_b_command',
+      'arguments' => array(
+        'code' => bt('The PHP code to evaluate.'),
+      ),
+      'aliases' => array('ev', 'eval'),
+      'bootstrap' => B_BOOTSTRAP_FULL,
+      'examples' => array(
+        'b php-eval \'$node = node_load(1); print $node->title;\'' => 'Loads node with nid 1 and then prints its title.',
+        'b php-eval "node_access_rebuild();"' => 'Rebuild node access permissions.',
+        'b php-eval "file_unmanaged_copy(\'$HOME/Pictures/image.jpg\', \'public://image.jpg\');"' => 'Copies a file whose path is determined by an environment\'s variable. Note the use of double quotes so the variable $HOME gets replaced by its value.',
+      ),
+    ),
+  );
+}
+
+/**
+ * Command callback: Execute specified PHP code.
+ */
+function php_eval_b_command($arguments, $options) {
+  try {
+    eval($arguments['code'] . ';');
```

Comment:
If we're going to add a semicolon to the end of the PHP code, let's first make sure there isn't one already...

Reply:
We don't need to. There is no harm in appending extra semicolons, even if superfluous. (That's also why the drush command does it this way as well.) And so it makes this code shorter and easier to understand.

Reply:
I know "we don't need to", but that doesn't mean we shouldn't. Any empty PHP comment doesn't do any harm, but it wouldn't be allowed to remain in a core PR because we have coding standards and (as much as possible) we try to do things 'the right way'.

```suggestion
    $code = $arguments['code'];
    $code .= (substr($code, -1) != ';') ? ';' : '';
    eval($code);
```

Reply:
That's still going to add an extra ';' if there's trailing whitespace in the executed string because ';' would no longer be the final character, but adding `trim()` to line 34 would take care of that.

If the final character is a '}' then no ';' would be needed, but this routine would still add one. For example, this would be a valid input, no terminal semicolon needed: `if (TRUE) { print("foo"); }`.

So, yes, we could trap that one, too. But it seems like trying to trap all the cases where an extra semicolon would be superfluous adds needlessly to the code, even if those are the only two possibilities (whitespace and '}'). Just tacking on an extra semicolon covers all use cases, doesn't generate any extra PHP evaluation, and is more readable because it's shorter.

So I'll ask: is this really important?

Reply:
I'll also point out a few more examples that could be problematic for determining whether a terminal semicolon is needed:

* `b ev 'print "foo"; // some comment'`    <-- not needed
* `b ev 'print "foo" /* some comment */'`    <-- needed
* `b ev "print 'foo' // some comment ; "$'\n'" /* more */"`   <-- needed
* `b ev "print 'foo' // some comment "$'\n'" ; /* more */"`   <-- not needed

By the time we've got code that can handle the presence or absence of whitespace, braces, and comments, we're well down the path of parsing the PHP...before handing it over to the PHP parser.

Whereas tacking on an extra ';' does the job, and has years of successful use in drush. Just seems like we might as well stick with the proven, tested, and familiar approach.

Reply:
Hmmm... I don't really like it, but fine. Let's leave it for now.